futures-util/src/future/maybe_done.rs

```diff
@@ -48,21 +47,25 @@ impl<Fut: Future> MaybeDone<Fut> {
     /// The output of this method will be [`Some`] if and only if the inner
     /// future has been completed and [`take_output`](MaybeDone::take_output)
     /// has not yet been called.
+    #[project]
     #[inline]
     pub fn output_mut(self: Pin<&mut Self>) -> Option<&mut Fut::Output> {
-        unsafe {
-            let this = self.get_unchecked_mut();
-            match this {
-                MaybeDone::Done(res) => Some(res),
-                _ => None,
-            }
+        #[project]
+        match self.project() {
+            MaybeDone::Done(res) => Some(res),
+            _ => None,
         }
     }
 
     /// Attempt to take the output of a `MaybeDone` without driving it
     /// towards completion.
     #[inline]
     pub fn take_output(self: Pin<&mut Self>) -> Option<Fut::Output> {
+        // Safety: we return immediately unless we are in the `Done`
+        // state, which does not have any pinning guarantees to uphold.
+        //
+        // Hopefully `pin_project` will support this safely soon:
+        // https://github.com/taiki-e/pin-project/issues/184
```

Comment:
cc https://github.com/taiki-e/pin-project/issues/184, in the public types, it is certainly not easy to use the way I said.

Reply:
Yeah this file is what lead me to that problem :)